driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,350 @@
+    fn current(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        let time_since_epoch = now.duration_since(SystemTime::UNIX_EPOCH)?;
+        Ok(Self(time_since_epoch.as_secs() / BATCH_DURATION.as_secs()))
+    }
+
+    fn currently_being_solved(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        Self::current(now).map(|batch_id| batch_id.prev())
+    }
+
+    fn order_collection_start_time(self) -> SystemTime {
+        SystemTime::UNIX_EPOCH + Duration::from_secs(self.0 * BATCH_DURATION.as_secs())
+    }
+
+    fn solve_start_time(self) -> SystemTime {
+        self.order_collection_start_time() + BATCH_DURATION
+    }
+
+    fn next(self) -> BatchId {
+        self.0.checked_add(1).map(BatchId).unwrap()
+    }
+
+    fn prev(self) -> BatchId {
+        self.0.checked_sub(1).map(BatchId).unwrap()
+    }
+}
+
+#[derive(Debug)]
+pub struct AuctionTimingConfiguration {
+    /// The offset from the start of a batch at which point we should start
+    /// solving.
+    pub target_start_solve_time: Duration,
+    /// The offset from the start of the batch at which point there is not
+    /// enough time left to attempt to solve.
+    pub latest_solve_attempt_time: Duration,
+}
+
+pub struct Scheduler<'a> {
+    driver: &'a mut dyn StableXDriver,
+    auction_timing_configuration: AuctionTimingConfiguration,
+    last_solved_batch: Option<BatchId>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum Action {
+    Solve(BatchId),
+    Sleep(Duration),
+}
+
+impl<'a> Scheduler<'a> {
+    pub fn new(
+        driver: &'a mut dyn StableXDriver,
+        auction_timing_configuration: AuctionTimingConfiguration,
+    ) -> Self {
+        assert!(auction_timing_configuration.latest_solve_attempt_time <= BATCH_DURATION);
+        assert!(
+            auction_timing_configuration.target_start_solve_time
+                < auction_timing_configuration.latest_solve_attempt_time
+        );
+        Self {
+            driver,
+            auction_timing_configuration,
+            last_solved_batch: None,
+        }
+    }
+
+    pub fn run_forever(&mut self) {
+        loop {
+            thread::sleep(self.run_forever_single_iteration(SystemTime::now()));
+        }
+    }
+
+    /// Returns how long to sleep before starting the next iteration.
+    fn run_forever_single_iteration(&mut self, now: SystemTime) -> Duration {
+        match self.determine_action(now) {
+            Ok(Action::Sleep(duration)) => {
+                info!("Sleeping {}s.", duration.as_secs());
+                duration
+            }
+            Ok(Action::Solve(batch_id)) => {
+                info!("Starting to solve batch {}.", batch_id.0);
+                match self.driver.run(batch_id.0.into()) {
+                    DriverResult::Ok => {
+                        info!("Batch {} solved without error.", batch_id.0);
+                        self.last_solved_batch.replace(batch_id);
+                        Duration::from_secs(0)
```

Comment:
ubernit: I would slightly prefer returning `None` in this case as its a little more clear that it should not sleep.

Reply:
Not very important to me but I don't agree because with that argument a 0 Duration never makes sense so I'll leave it like this.

Reply:
He issue is that `thread::sleep(0)` sometimes behaves like a `yield` (I would imagine that it does in Rust as well but I’m not sure). That means sleeping for 0 and not sleeping at all may have slightly different semantics and from the code I don’t think yielding the thread is intended.

Reply:
Also, as it is right now, we would usually sleep for 0 and then sleep for the desired duration.

Reply:
I am leaving this in because it is benign and it will change when we merge the each batch in own thread anyway.